### PR TITLE
Add test-infra scripts

### PR DESCRIPTION
**Pull request by ArangoGutierrez**
initial PR to add NFD-Operator to kubernetes/test-infra CI platform
closes: #1 
closes: #9 

Signed-off-by: Carlos Eduardo Arango Gutierrez <carangog@redhat.com>

### CODE REVIEW

**Comment by ArangoGutierrez** [comment]
This is basically a copy/paste from https://github.com/kubernetes-sigs/node-feature-discovery to walk towards adding this repo to test-infra CI 
thanks for the effort there @marquiz

**Comment by ArangoGutierrez** [comment]
test-infra PR : image-pushing: add config for nfd-operator [#19141](https://github.com/kubernetes/test-infra/pull/19141)

File: Makefile [hunk 1]
```diff
@@ -1,7 +1,24 @@
...
-TAG            ?= $(shell git describe --always --dirty)
-IMAGE          ?= $(REGISTRY)/$(ORG)/node-feature-discovery-operator:$(TAG)
+.PHONY: all test
+.FORCE:
+
+GO_CMD := go
+GO_FMT := gofmt
+
+IMAGE_BUILD_CMD ?= docker build
+IMAGE_BUILD_EXTRA_OPTS :=
```
Comment (marquiz) [inline Makefile:8]: Could use `?=` for all "customization variables" in the way you proposed for NFD. WDYT(?)

I think that would include all variables from here to `IMAGE_EXTRA_TAGS`, excluding `VERSION`, `IMAGE_NAME` and possibly `IMAGE_REPO`.

File: Makefile [hunk 2]
```diff
@@ -1,7 +1,24 @@
-REGISTRY       ?= quay.io
-ORG            ?= kubernetes_incubator
-TAG            ?= $(shell git describe --always --dirty)
-IMAGE          ?= $(REGISTRY)/$(ORG)/node-feature-discovery-operator:$(TAG)
+.PHONY: all test
```
Comment (marquiz) [inline Makefile:1]: Don't we have duplicat .PHONY entries somewhere later??

File: Makefile [hunk 3]
```diff
@@ -1,7 +1,24 @@
...
+.FORCE:
+
+GO_CMD := go
+GO_FMT := gofmt
```
Comment (marquiz) [inline Makefile:5]: Not used?? Use in the `verify-gofmt` target?

File: scripts/test-infra/build-image.sh [hunk 4]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash -e
+
+make image -e
```
Comment (marquiz) [inline scripts/test-infra/build-image.sh:3]: If the variable initializations in Makefile is changed to `?=` we could simply use `make image` without `-e`

File: scripts/test-infra/build.sh [hunk 5]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash -e
+
+make build -e
```
Comment (marquiz) [inline scripts/test-infra/build.sh:3]: If the variable initializations in Makefile is changed to ?= we could simply use make image without -e

File: scripts/test-infra/push-image.sh [hunk 6]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash -e
+
+# Strip 'vYYYYMMDD-' from the variable in order to get a reproducible
+# version and container image tag
+if [ -n "$_GIT_TAG" ]; then
+    export VERSION=${_GIT_TAG:10}
+fi
+
+# Authenticate in order to be able to push images
+gcloud auth configure-docker
```
Comment (marquiz) [inline scripts/test-infra/push-image.sh:10]: I'm not actually sure we need this. Could be dropped (from NFD, too)

File: scripts/test-infra/test-e2e.sh [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+        exit 1
+    fi
+    sleep 60
+
+    i=$(( $i + 1 ))
+done
+
+
+# Configure environment and run tests
+make e2e-test -e
```
Comment (marquiz) [inline scripts/test-infra/test-e2e.sh:33]: I don't think we have any e2e tests. Let's add this script when we know what/how/where to test

File: scripts/test-infra/verify.sh [hunk 8]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash -e
+
+# Install deps
+curl -sfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh| sh -s -- -b $(go env GOPATH)/bin v1.27.0
```
Comment (marquiz) [inline scripts/test-infra/verify.sh:4]: How about using version 1.30.0, á la https://github.com/kubernetes-sigs/node-feature-discovery/pull/349

File: scripts/test-infra/verify.sh [hunk 9]
```diff
@@ -0,0 +1,9 @@
...
+export PATH=$PATH:$(go env GOPATH)/bin
+
+# Run verify steps
+make gofmt-verify -e
+make ci-lint -e
```
Comment (marquiz) [inline scripts/test-infra/verify.sh:9]: Here we could also drop `-e`, I think

We don't have these targets in the Makefile? Just do `make verify` and add golangci-lint as a new make target that `verify` depends on?

**Review by marquiz — CHANGES_REQUESTED** [review]
Thanks @ArangoGutierrez ! I think this is a good start, but a bit too many copy-paster errors and inconsistencies 😉  I surely didn't spot all of them

File: Makefile [hunk 10]
```diff
@@ -20,6 +37,18 @@ DOCKERFILE=Dockerfile
...
+# Take CONTAINER_HOSTMOUNT_PREFIX from HOSTMOUNT_PREFIX if only the latter is specified
+ifdef HOSTMOUNT_PREFIX
+    CONTAINER_HOSTMOUNT_PREFIX := $(HOSTMOUNT_PREFIX)
+else
+    CONTAINER_HOSTMOUNT_PREFIX := /host-
+endif
+HOSTMOUNT_PREFIX := /
+
+KUBECONFIG :=
+E2E_TEST_CONFIG :=
```
[the remark this answers is not in the document]
Comment (ArangoGutierrez) [reply Makefile:50]: Yeahhh totally :( fixing

**Comment by ArangoGutierrez** [comment]
Thanks for all the checks @marquiz , as you said lot's of copy/paste errors, now fixed.

File: Makefile [hunk 11]
```diff
@@ -1,7 +1,17 @@
-REGISTRY       ?= quay.io
-ORG            ?= kubernetes_incubator
-TAG            ?= $(shell git describe --always --dirty)
-IMAGE          ?= $(REGISTRY)/$(ORG)/node-feature-discovery-operator:$(TAG)
+IMAGE_BUILD_CMD ?= docker build
+IMAGE_BUILD_EXTRA_OPTS ?=
+IMAGE_PUSH_CMD ?= docker push
```
Comment (marquiz) [inline Makefile:3]: This is unused(?) You might want to add a `push` or `push-images` target using this as you do `make push` later in `push-image.sh`

Comment (ArangoGutierrez) [reply Makefile:3]: good catch, done!

File: Makefile [hunk 12]
```diff
@@ -1,7 +1,17 @@
...
+
+VERSION := $(shell git describe --tags --dirty --always)
+
+IMAGE_REGISTRY := k8s.gcr.io/nfd
+IMAGE_NAME := node-feature-discovery-operator
+IMAGE_TAG_NAME := $(VERSION)
+IMAGE_EXTRA_TAG_NAMES :=
+IMAGE_REPO := $(IMAGE_REGISTRY)/$(IMAGE_NAME)
+IMAGE_TAG := $(IMAGE_REPO):$(IMAGE_TAG_NAME)
+IMAGE_EXTRA_TAGS := $(foreach tag,$(IMAGE_EXTRA_TAG_NAMES),$(IMAGE_REPO):$(tag))
```
Comment (marquiz) [inline Makefile:13]: I think we could use `?=` for most (if not all) of these – probably `IMAGE_REPO` and `IMAGE_NAME` could be considered "internal" variables and initialized with `:=`. That way, we could drop the `-e` make option from the test-infra scripts.

**Review by marquiz — CHANGES_REQUESTED** [review]
I noticed that the permissions of `push-image.sh` are incorrect, need to add `+x`.

Please test/verify the Makefile and the test-infra scripts, e.g. by running against e.g. a personal container registry, and report back.

**Comment by ArangoGutierrez** [comment]
> I noticed that the permissions of `push-image.sh` are incorrect, need to add `+x`.
> 
> Please test/verify the Makefile and the test-infra scripts, e.g. by running against e.g. a personal container registry, and report back.
Fixed

Tested the tst_infra scripts

Tested against  **quay.io/eduardoarango/node-feature-discovery-operator:40a8af9-dirty**

File: Makefile (hunk 12 above)
Comment (ArangoGutierrez) [reply Makefile]: agree , mostly for testing, you want to override those for a personal repo

File: scripts/test-infra/verify.sh [hunk 13]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash -e
+
+# Run verify steps
+make verify
```
Comment (marquiz) [inline scripts/test-infra/verify.sh:4]: Missing `verify` target from the makefile(?)

File: Makefile [hunk 14]
```diff
@@ -1,7 +1,17 @@
...
-TAG            ?= $(shell git describe --always --dirty)
-IMAGE          ?= $(REGISTRY)/$(ORG)/node-feature-discovery-operator:$(TAG)
+IMAGE_BUILD_CMD ?= docker build
+IMAGE_BUILD_EXTRA_OPTS ?=
+IMAGE_PUSH_CMD ?= docker push
+
+VERSION := $(shell git describe --tags --dirty --always)
+
+IMAGE_REGISTRY ?= k8s.gcr.io/nfd
+IMAGE_NAME ?= node-feature-discovery-operator
```
Comment (marquiz) [inline Makefile:8]: Nit: I think IMAGE_NAME should be initialized with `:=`

**Review by marquiz — COMMENTED** [review]
Two new small comments. There are still few open comments from the previous reviews

File: scripts/test-infra/verify.sh (hunk 13 above)
Comment (ArangoGutierrez) [reply scripts/test-infra/verify.sh:4]:
``` 
 64 verify: verify-gofmt
 65 
 66 verify-gofmt:
 67         @./scripts/verify-gofmt.sh
```

File: Makefile (hunk 14 above)
Comment (ArangoGutierrez) [reply Makefile:8]: fixed

File: scripts/test-infra/verify.sh (hunk 13 above)
Comment (marquiz) [reply scripts/test-infra/verify.sh:4]: Woops, sorry 😬

Comment (ArangoGutierrez) [reply scripts/test-infra/verify.sh:4]: /honk

File: Makefile (hunk 14 above)
Comment (marquiz) [reply Makefile:8]: Is it??

File: scripts/test-infra/verify.sh (hunk 8 above)
Comment (marquiz) [reply scripts/test-infra/verify.sh]: Hmm, you dropped golangci-lint. What was the reason for that?

**Comment by marquiz** [comment]
Thanks @ArangoGutierrez. Still few unresolved conversation, just some minor things

File: Makefile (hunk 14 above)
Comment (ArangoGutierrez) [reply Makefile:8]: No? damm didn't got into the `git add` one sec

Comment (ArangoGutierrez) [reply Makefile]: how about now?

File: scripts/test-infra/build.sh (hunk 5 above)
Comment (marquiz) [reply scripts/test-infra/build.sh:3]: What about this?

File: scripts/test-infra/push-image.sh (hunk 6 above)
Comment (marquiz) [reply scripts/test-infra/push-image.sh:10]: and this?

File: scripts/test-infra/verify.sh (hunk 8 above)
Comment (marquiz) [reply scripts/test-infra/verify.sh]: and this?

**Review by marquiz — COMMENTED** [review]
...still 3 open conversations...

File: scripts/test-infra/verify.sh (hunk 8 above)
Comment (ArangoGutierrez) [reply scripts/test-infra/verify.sh:4]: > ...still 3 open conversations...



> ...still 3 open conversations...

Thanks for pointing me out. 
updated now

File: scripts/test-infra/push-image.sh [hunk 15]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash -e
+
+# Strip 'vYYYYMMDD-' from the variable in order to get a reproducible
+# version and container image tag
+if [ -n "$_GIT_TAG" ]; then
+    export VERSION=${_GIT_TAG:10}
+fi
+
+make image 
+make push
```
Comment (marquiz) [inline scripts/test-infra/push-image.sh:10]: We need to override version. Better to do it explicitly than with `-e` I think
```
make image VERSION="$VERSION"
make push VERSION="$VERSION"
```

**Review by marquiz — CHANGES_REQUESTED** [review]
There is a typo in the commit message: s/iverride/override/  😉. Maybe you could just squash the commits

**Review by marquiz — APPROVED** [review]
Thanks @ArangoGutierrez 
/lgtm

**Pull request merged**